gcc/rust/typecheck/rust-tyty-rules.h

```diff
@@ -242,6 +242,11 @@ class BaseRules : public TyVisitor
 		   type.as_string ().c_str ());
   }
 
+  virtual void visit (NeverType &type) override
+  {
+    resolved = get_base ()->clone ();
+  }
```

Comment:
> And what's more, assigning `<integer>` to `!` is illegal and assigning `!` to `<integer>` is legal. At present, it seems that we do not distinguish between `A->unify(B)` and `B->unify(A)`.

Wouldn't emitting an error here solve this? 

If `A=!` and `B=?`, then this is the direction `B->unify(A)` where you have a type `B` and try to make it work with the `!` type, aka assigning B to A (?). This is only allowed if `B==A==!`, which is covered by the `NeverRules`. 

On the other hand, the `A->unify(B)` is handled in the `NeverRules`, which means that you try to make `A` work with `B`, aka assigning A to B, which is always allowed.

Reply:
I don't think it helps. My opinion is that we must distinguish `unify` and `assign` explicitly. For example, we do need unify `!` with `<integer>` as well as unifying `<integer>` with `!` type. Consider the following two code segments.
```rust
let a = if x > y {
    return -1; // ! type
} else {
    1 // <integer> type
}; // `!`->unify(`<integer>`);
```
```rust
let a = if x > y {
   1 // <integer> type
} else {
    return -1; // ! type
}; // `<integer>`->unify(`!`);
````

Reply:
Ah yeah, I tried to think about a situation where the other direction also works. Should have thought about it longer :smile: Nvm then. This has to be addressed, but I would say this is too big for this PR. I suggest to add a `FIXME` here as well. Maybe also add a `XPASS` test that does `let a: ! = 123;` (not sure if GCC has xpass tests though).

Reply:
Yeah, I totally agree with this.